shell/utils/resourceWatcher.js

```diff
@@ -0,0 +1,212 @@
+    } = WATCHSTATUSES;
+
+    if (this.isConnected()) {
+      if (watchesArray.length === 0) {
+        this.disconnect();
+      } else {
+        watchesArray
+          .forEach((watch) => {
+            const {
+              resourceType, id, namespace, selector, resourceVersion, status, skipResourceVersion
+            } = watch;
+            const watchObject = {
+              resourceType,
+              id,
+              namespace,
+              selector
+            };
+            const watchKey = keyForSubscribe(watchObject);
+
+            if (status === PENDING) {
+              this.send(JSON.stringify({
+                ...watchObject,
+                resourceVersion: !skipResourceVersion ? resourceVersion : undefined
+              }));
+              this.watches[watchKey].status = REQUESTED;
+            } else if (watch.status === CANCELLED) {
+              this.send(JSON.stringify({
+                ...watchObject,
+                stop: true
+              }));
+              this.watches[watchKey].status = REMOVED;
+            }
+          });
+      }
+    } else if (watchesArray.length > 0) {
+      watchesArray
+        .forEach((watch) => {
+          const { status } = watch;
+          const watchKey = keyForSubscribe(watch);
+
+          if ([PENDING, REQUESTED, WATCHING, REFRESHING].includes(status)) {
+            this.watch(watchKey);
+          } else {
+            this.watch[watchKey].status = REMOVED;
+          }
+        });
+
+      this.connect(this.connectionMetadata);
+    }
+  }
+
+  cleanupWatches() {
+    // get rid of all of the removed watches
+    Object.values(this.watches)
+      .filter(watch => watch.status === WATCHSTATUSES.REMOVED)
+      .forEach((watch) => {
+        delete this.watches[keyForSubscribe(watch)];
+      });
+  }
+
+  _onmessage(event) {
+    const {
+      REQUESTED, WATCHING, STOPPED, CANCELLED, REMOVED
+    } = WATCHSTATUSES;
+
+    const {
+      name: eventName, resourceType, data: { type }, id, namespace, selector, data
+    } = JSON.parse(event.data);
+    const watchKey = keyForSubscribe({
+      resourceType,
+      type,
+      id,
+      namespace,
+      selector
+    });
+
+    if (eventName === 'resource.start' && this.watches?.[watchKey]?.status === REQUESTED) {
+      this.watches[watchKey].status = WATCHING;
+    } else if (eventName === 'resource.stop' && this.watches?.[watchKey] && ![CANCELLED, REMOVED].includes(this.watches?.[watchKey]?.status)) {
+      this.watches[watchKey].status = STOPPED;
+      delete this.watches[watchKey].resourceVersion;
+      delete this.watches[watchKey].resourceVersionTime;
+      this.watch(watchKey);
+    } else if (eventName === 'resource.error') {
+      const err = data?.error?.toLowerCase();
+
+      if ( this.watches[watchKey] && err.includes('watch not allowed') ) {
+        this.watches[watchKey].error = { type: resourceType, reason: NO_WATCH };
```

Comment:
Are the errors cleared anywhere?

Reply:
it'll be cleared upon a successful watch when I rewrite the watchObject